Add image classification examples

Honry opened the pull request:
Supports MobileNet and SqueezeNet with both 'NCHW' and 'NHWC' layouts.

Use MobileNet V2, SqueezeNet 1.1 from ONNX and MobileNet V2,
SqueezeNet 1.0 from TFLite as model baselines.

Centralize some common resources from style_transfer to 'common' folder BTW.

Fixed #5, #6

Honry commented:
@huningxin, PTAL, thanks!

huningxin commented:
Super. Thanks @Honry ! Do you have a preview link available?

Honry commented:
@huningxin, here's the preview: https://honry.github.io/webnn-samples/image_classification/

huningxin commented:
Thanks for sharing the preview link, @Honry ! It looks beautiful.

Some high-level comments:
1. The weights downloading takes time, could the progress be displayed on the UI? so user knows the status.
2. When I pick another image, it seems to download and build the model again, even without changing the model. Could it just compute?
3. Similar behavior when I just switch to camera view. I suppose it could reuse the compiled graph.
4. There is a warning saying "Cannot use 'in' operator to search for 'dispose' in null" when I pick another image.

huningxin commented on common/utils.js:
```diff
@@ -28,15 +28,82 @@ export async function buildConstantByNpy(builder, url) {
+    const buffer = await b.buffer();
+    dimensions = b.shape;
+    typedArray = buffer.values;
+    tf.dispose();
```
I understand you need to dispose both `a` and `b`. However, probably it's good to wrap these code section into [`tf.tidy`](https://js.tensorflow.org/api/latest/#tidy).

Honry commented:
@huningxin, many thanks for your comments! I've addressed most of the comments except the first one.

> The weights downloading takes time, could the progress be displayed on the UI? so user knows the status.

Since the weights downloading are wrapped in the `load()` method, we don't know the progress "out of the box", one way I can imagine is to hack each weight's loading to track the progress, but this would break the style of `load()` method and is not friendly for user. So I would suggest to skip this feature and see if there's better solution in future.

huningxin commented on common/utils.js:
```diff
@@ -28,15 +28,82 @@ export async function buildConstantByNpy(builder, url) {
   }
+  // TODO(Wanming): This is a workaround to transpose 'ihwo' to 'hwio',
+  // and will be removed once 'ihwo' filterLayout is supported.
+  if (isDepthwise) {
```
If it is a workaround for a specific sample, I'd prefer to move it into the sample. You can add a `transpose` op in the graph after create the constant.

Honry replied:
The `transpose` op in WebNN only supports transpose a `MLOperand`, that's why I have to do this workaround in `buildConstantByNpy`.

Honry replied:
Good suggestion! It's my misunderstanding the `transpose` op, I thought it could not transpose a `Constant`.

huningxin commented on common/utils.js:
```diff
@@ -28,15 +28,82 @@ export async function buildConstantByNpy(builder, url) {
+/**
+ * This method is used to covert input element to tensor data.
+ * @param {Object} inputElement, an object of HTML [<img> | <video>] element.
+ * @param {!Object<string, *>} options, an object of options to process
```
could you specify which options are supported?

huningxin commented on image_classification/main.js:
```diff
@@ -0,0 +1,275 @@
+  mean: [0.485, 0.456, 0.406],
+  std: [0.229, 0.224, 0.225],
+  norm: true,
+  nchwFlag: true,
```
`inputLayout`?

huningxin commented on image_classification/mobilenet_nchw.js:
```diff
@@ -0,0 +1,132 @@
+
+import {buildConstantByNpy} from '../common/utils.js';
+
+// MobileNet V2 baseline model with nchw layout
```
remove "baseline", add "input" before layout.

huningxin commented on image_classification/mobilenet_nchw.js:
```diff
@@ -0,0 +1,132 @@
+import {buildConstantByNpy} from '../common/utils.js';
+
+// MobileNet V2 baseline model with nchw layout
+export class MobileNetNchw {
```
`MobileNetV2Nchw

huningxin commented on image_classification/main.js:
```diff
@@ -0,0 +1,275 @@
+  nchwFlag: true,
+  labelUrl: './labels/labels1000.txt',
+  inputDimensions: [1, 3, 224, 224],
+};
```
I think the options should be part of model, please move it respectively.

huningxin commented on image_classification/mobilenet_nchw.js:
```diff
@@ -0,0 +1,132 @@
+    const conv = this.builder_.add(
+        this.builder_.conv2d(input, weights, options),
+        this.builder_.reshape(bias, [1, -1, 1, 1]));
+    if (shouldClip) {
```
please name it `relu6`, that is used by the original model, we just use `clamp` to implement `relu6`.

huningxin commented on image_classification/mobilenet_nchw.js:
```diff
@@ -0,0 +1,132 @@
+    return this.builder_.gemm(input, weights, options);
+  }
+
+  async buildFire_(
```
`Fire` is not used in MobileNetV2 (it's used for SqueezeNet). According to the [paper](https://arxiv.org/pdf/1801.04381.pdf), we should build `Bottleneck` block.  And there are two types of bottleneck block, as it shows in the diagram:

![image](https://user-images.githubusercontent.com/1005673/114963686-df672680-9e9f-11eb-8958-42e5af966440.png)

And I'd suggest to create a helper to build depthwise conv explicitly. The helper could create grouped conv with groups == input channels.

huningxin commented:
> one way I can imagine is to hack each weight's loading to track the progress, but this would break the style of `load()` method and is not friendly for user. So I would suggest to skip this feature and see if there's better solution in future.

sounds good. I agree the friendly code is more important in the context of WebNN API example.

Honry commented:
@huningxin, thanks for your further comments, really helpful! PTAL again, thanks!

huningxin commented on image_classification/mobilenet_nhwc.js:
```diff
@@ -0,0 +1,130 @@
+/* eslint max-len: ["error", {"code": 120}] */
+
+// MobileNet V2 model with 'nhwc' input layout
+export class MobileNetV2Nhwc {
```
Please align to `MobileNetV2Nchw` implementation.

huningxin commented on image_classification/mobilenet_nhwc.js:
```diff
@@ -0,0 +1,130 @@
+        conv0, '238', 'expanded_conv_depthwise_depthwise', true, {autoPad, groups: 32, filterLayout: 'hwio'});
+    const conv2 = await this.buildConv_(
+        conv1, '167', 'expanded_conv_project_Conv2D', false, {autoPad, filterLayout});
+    const fire0 = await this.buildBottleneck_(conv2, ['165', '99', '73'], '1', {strides, groups: 96});
```
rename `fire0` to `bottleneck0`

huningxin commented on image_classification/mobilenet_nhwc.js:
```diff
@@ -0,0 +1,130 @@
+        conv1, '167', 'expanded_conv_project_Conv2D', false, {autoPad, filterLayout});
+    const fire0 = await this.buildBottleneck_(conv2, ['165', '99', '73'], '1', {strides, groups: 96});
+    const fire1 = await this.buildBottleneck_(fire0, ['3', '119', '115'], '2', {groups: 144});
+    const add0 = this.builder_.add(fire0, fire1);
```
combine this `add` into `buildLinearBottleneck_` `shortcut` parameter.

Honry commented:
@huningxin, thanks again! All comments fixed, PTAL, thanks!

huningxin commented on common/.eslintrc.js:
```diff
@@ -3,5 +3,6 @@ module.exports = {
   globals: {
     'BigInt64Array': 'readonly',
     'BigUint64Array': 'readonly',
+    'tf': 'readonly',
```
do we still need this one?

Honry replied:
Thanks, I've removed it.

huningxin reviewed (approved):
Nice work. LGTM with a nit. Please fix it and then we can merge. Thanks!

huningxin commented:
@Honry @BruceDai , there are many weight files (*.npy) in this PR. As the webnn-samples and tests in webnn-polyfill (probably later webnn-native) will share them, we may need to find a place to host them separately. Probably a new repo under webmachinelearning, say github.com/webmachinelearning/test-data?

Any thoughts?

also cc @anssiko for comments.

Honry commented:
@huningxin, good idea! Then we can make it as a git submodule for both webnn-samples and webnn-polyfill.

BruceDai commented:
@huningxin Yes, adding new repo is a good suggestion. How about this name github.com/webmachinelearning/models_zoo?

huningxin commented:
> How about this name github.com/webmachinelearning/models_zoo?

@BruceDai , thanks for the suggestion. However I suppose we will not host any models (like .onnx or .tflite files) there. And webnn-polyfill and webnn-native may share the test data other than model weights.

BruceDai commented:
@huningxin Thanks. That sounds good.

anssiko commented:
I support @huningxin’s proposal for a new repo `webmachinelearning/test-data` to host reusable .npy weight files and other test data.

@huningxin feel free to create the repo and please add an appropriate readme to this new repo to note projects that are expected to reuse these assets.

huningxin commented:
> feel free to create the repo and please add an appropriate readme to this new repo to note projects that are expected to reuse these assets.

done: https://github.com/webmachinelearning/test-data

@Honry , could you please extract the npy files in this PR and include them in a new PR to test-data repo.

We may create folder like 'test-data/models/mobilenetv2_nchw/`. I would suggest to post a README file for each model that includes the info of the origin and license.

Once the weights of mobilenet and squeezenet are available in the test-data repo, please also update the model tests in polyfill to use them.

Further, we may need to move LeNet and NSNet2 weights to test-data.

Honry commented:
@huningxin, sure, I will make it as a git submodule to webnn-samples and webnn-polyfill.

Honry commented:
@huningxin, I've removed the weight files in this PR. PTAL, thanks!

huningxin commented:
LGTM. Great work, thanks again @Honry !
